### PR TITLE
Harden the PathString converter

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/aspnet/HttpAbstractions/issues/848 https://github.com/aspnet/HttpAbstractions/issues/690

@Tratcher

### CODE REVIEW

**Comment by davidfowl** [comment]
Isn't this in preview2?

**Comment by HaoK** [comment]
Yeah this is just cleanup, but I can make this change in preview 2 if we want

File: src/Microsoft.AspNetCore.Http.Abstractions/PathString.cs [hunk 1]
```diff
@@ -459,9 +459,20 @@ public override int GetHashCode()
...
+            => sourceType == typeof(string) 
+            ? true 
+            : base.CanConvertFrom(context, sourceType);
+
         public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
-        {
-            return new PathString((string)value);
-        }
+            => value is string 
+            ? new PathString((string)value)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/PathString.cs]: This replicates the asymmitry bug from the implicit converters:
https://github.com/aspnet/HttpAbstractions/issues/690

string -> PathString should use FromUriComponent(string), not new PathString(string)

We should change them both at the same time. @davidfowl

Comment (HaoK) [inline src/Microsoft.AspNetCore.Http.Abstractions/PathString.cs]: Let me see if I can unify the code a bit to make that change easier

**Comment by HaoK** [comment]
So I added a new 
```
internal static PathString ConvertFromString(string s) => string.IsNullOrEmpty(s) ? new PathString(s) : FromUriComponent(s);
```

That both the implicit op and ConvertFrom call, but it wasn't obvious to me exactly how to write a test verifying that the round tripping issue was fixed.

**Comment by Tratcher** [comment]
The most obvious things that are broken by the asymmetry are the equality operators. E.g. Equals, StartsWithSegments, etc..
Here's a repro:
```
            PathString p1 = "/?";
            String s1 = p1; // Becomes "/%3F"
            PathString p2 = s1; // Stays "/%3F"
            Console.WriteLine(p1.Equals(p2)); // false
```

**Comment by HaoK** [comment]
Added a test verifying the equality now work

**Comment by HaoK** [comment]
Now also fixes: https://github.com/aspnet/HttpAbstractions/issues/690

File: test/Microsoft.AspNetCore.Http.Abstractions.Tests/PathStringTests.cs [hunk 2]
```diff
@@ -208,11 +208,28 @@ public void ToUriComponentEscapeCorrectly(string category, string input, string
...
+            Assert.False(converter.CanConvertTo(typeof(int)));
+            Assert.False(converter.CanConvertTo(typeof(bool)));
         }
+
+        [Fact]
+        public void PathStringStaysEqualAfterAssignments()
+        {
+            PathString p1 = "/?";
+            string s1 = p1; // Becomes "/%3F"
+            PathString p2 = s1; // Stays "/%3F"
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Http.Abstractions.Tests/PathStringTests.cs]: `Stays "/%3F"` is wrong after your fix

**Comment by Tratcher** [comment]
*ping*

**Comment by HaoK** [comment]
Yeah I'll merge this later today

**Comment by HaoK** [comment]
12f89f66a687bdb1c1b683b0d7abd2732c05eb1a

**Pull request closed without merging**